ide/src/extension.ts

```diff
@@ -1,28 +1,48 @@
 import * as vscode from "vscode";
 import _ from "lodash";
-import { CallFlowistry, log, show_error } from "./vsc_utils";
+import {
+  CallFlowistry,
+  FlowistryErrorDocument,
+  last_error,
+  log,
+  show_error,
+} from "./vsc_utils";
 
 import { decompose } from "./decompose";
-import { focus, focus_mark, focus_unmark, focus_select } from "./focus";
+import { FocusMode } from "./focus";
 import { setup } from "./setup";
 
 import "./app.scss";
 
+export let flowistry_status_bar_item: vscode.StatusBarItem;
+export const tdcp = new FlowistryErrorDocument();
+
 export async function activate(context: vscode.ExtensionContext) {
   log("Activating...");
 
   try {
+    flowistry_status_bar_item = vscode.window.createStatusBarItem();
+    context.subscriptions.push(flowistry_status_bar_item);
+    flowistry_status_bar_item.show();
+
+    context.subscriptions.push(
+      vscode.workspace.registerTextDocumentContentProvider("flowistry", tdcp)
+    );
+
     let call_flowistry = await setup(context);
     if (call_flowistry === null) {
       return;
     }
 
+    let focus = new FocusMode(call_flowistry, flowistry_status_bar_item);
+
     let commands: [string, (_f: CallFlowistry) => void][] = [
-      ["focus", focus],
-      ["focus_mark", focus_mark],
-      ["focus_unmark", focus_unmark],
-      ["focus_select", focus_select],
+      ["focus", focus.focus.bind(focus)],
+      ["focus_mark", focus.focus_mark.bind(focus)],
+      ["focus_unmark", focus.focus_unmark.bind(focus)],
+      ["focus_select", focus.focus_select.bind(focus)],
```

Comment:
Each `focus.____` method loses its `this` context when registered as a command ([happens when passed to a HOF](https://eliux.github.io/javascript/common-errors/why-this-gets-undefined-inside-methods-in-javascript/)). Need to explicitly bind `focus` to access `this`.

Reply:
If you define the class methods as arrow functions, you won't need to call `.bind(...)`.